replace_text.js

```diff
@@ -1,13 +1,62 @@
+function rot13(str) {
+  return str.split("").map(function(c) {
+    var lowerCharCode = c.toLowerCase().charCodeAt(0);
+
+    if (lowerCharCode < 97 || lowerCharCode > 122) return c;
+
+    var plain = String.fromCharCode((lowerCharCode - 97 + 13) % 26 + 97);
+    return c.toUpperCase() === c ? plain.toUpperCase() : plain;
+
+  }).join("");
+}
+
+
 if (document.activeElement.nodeName === "TEXTAREA" || document.activeElement.nodeName === "INPUT") {
   var ta = document.activeElement;
-  ta.value = ta.value.slice(0, ta.selectionStart) + atob(plainText) + ta.value.slice(ta.selectionEnd, ta.length);
+  var start = ta.selectionStart;
+  var end = ta.selectionEnd;
+
+  ta.value = ta.value.slice(0, start) + rot13(ta.value.slice(start, end)) + ta.value.slice(end);
+
+  // Set selection back to its original position
+  ta.selectionStart = start;
+  ta.selectionEnd = end;
 }
 else {
   var sel = window.getSelection();
-  var range;
-  if (sel.rangeCount) {
-    range = sel.getRangeAt(0);
-    range.deleteContents();
-    range.insertNode(document.createTextNode(atob(plainText)));
+  for (var r = 0; r < sel.rangeCount; ++r) {
+    var range = sel.getRangeAt(0);
+    var startOffset = range.startOffset;
+    var endOffset = range.endOffset;
+
+    var selectedTextNodes = document.createNodeIterator(
+      range.commonAncestorContainer, NodeFilter.SHOW_TEXT, function(n) {
+        // we don't want to change <style> and <script> elements
+        var parent = n.parentNode.nodeName.toLowerCase();
+        if (parent === "style" || parent === "script")
+          return NodeFilter.FILTER_REJECT;
+
+        return sel.containsNode(n, true) ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_REJECT;
+      }
+    );
+
+    var node;
+    while (node = selectedTextNodes.nextNode()) {
+      // Work around a bug(?) in containsNode that unexpectedly returns true for
+      // the first text node in the element AFTER the selection
+      if(range.endContainer !== range.startContainer &&
+         range.endContainer.nodeType !== Node.TEXT_NODE &&
```

Comment:
Do you have a way to consistently repro this? I was messing around with the script and couldn't get it to happen. I wonder if it's an interaction/bug with setting partialContainment to true in `containsNode` up above.

Reply:
In particular I noticed that happening when I triple-clicked on some `<p>` text to select the entire paragraph. I suspect that internally it was counting all the elements _after_ that paragraph as part of the selection as well, all the way up until the start of the next actual text. This `if` was kind of a hacky workaround but it seemed to fix the problem at least in that specific case.

The value I gave to partialContainment didn't seem to affect the results either way in any situation that I tried, so either that isn't supported by the version of Chrome I was running, or I wasn't trying the right things. I am not primarily a web dev person so I'm not really sure what was going on here.

Reply:
Yeah, it wouldn't surprise me if a triple-click created selection is a bit wonky. It's also interesting that the range's endContainer isn't a TEXT_NODE in that case. But yeah, I'm not sure either. A mystery for another day :smile: